SmartDeviceLink/SDLUploadFileOperation.m

```diff
@@ -64,99 +69,165 @@ - (void)sdl_sendPutFiles:(NSArray<SDLPutFile *> *)putFiles withCompletion:(SDLFi
-            putFile.length = @(mtuSize);
-        }
+/**
+ Gets the size of the data to be sent in a packet. Packet size can not be greater than the max MTU size allowed by the SDL Core.
+
+ @param currentOffset The position in the file where to start reading data
+ @param fileSize The size of the file
+ @param mtuSize The maximum packet size allowed
+ @return The size of the data to be sent in the packet.
+ */
+- (NSUInteger)sdl_getDataSizeForOffset:(NSUInteger)currentOffset fileSize:(unsigned long long)fileSize mtuSize:(NSUInteger)mtuSize {
+    NSInteger dataSize = 0;
+    NSUInteger fileSizeRemaining = (NSUInteger)(fileSize - currentOffset);
+    if (fileSizeRemaining < mtuSize) {
+        dataSize = fileSizeRemaining;
+    } else {
+        dataSize = mtuSize;
+    }
+    return dataSize;
+}
 
-        // Place the data and set the new offset
-        if (currentOffset == 0 && (mtuSize < [putFile.length unsignedIntegerValue])) {
-            putFile.bulkData = [fileData subdataWithRange:NSMakeRange(currentOffset, mtuSize)];
-            currentOffset = mtuSize;
-        } else {
-            putFile.bulkData = [fileData subdataWithRange:NSMakeRange(currentOffset, [putFile.length unsignedIntegerValue])];
-            currentOffset += [putFile.length unsignedIntegerValue];
-        }
+/**
+ Reads a chunk of data from a socket.
 
-        [putFiles addObject:putFile];
+ @param size The amount of data to read from the input stream
+ @param inputStream The socket from which to read the data
+ @return The data read from the socket
+ */
+- (nullable NSData *)sdl_getDataChunkWithSize:(NSInteger)size inputStream:(NSInputStream *)inputStream {
+    if (size <= 0) {
+        return  nil;
+    }
+    
+    Byte buffer[size];
+    NSInteger bytesRead = [inputStream read:buffer maxLength:size];
+    if (bytesRead) {
+        NSData *dataChunk = [[NSData alloc] initWithBytes:(const void*)buffer length:size];
+        return dataChunk;
+    } else {
+        // TODO: return a custom error?
+        return nil;
     }
-
-    return putFiles;
 }
 
+/**
+ One of the responses returned by the SDL Core will contain the correct remaining free storage size on the SDL Core. Since communication with the SDL Core is asynchronous, there is no way to predict which response contains the correct bytes available other than to watch for the largest correlation id, since that will be the last response sent by the SDL Core.
 
-#pragma mark Property Overrides
+ @param request The newest response returned by the SDL Core for a putfile
+ @param highestCorrelationIDReceived The largest currently received correlation id
+ @return Whether or not the newest request contains the highest correlationId
+ */
+- (Boolean)sdl_newHighestCorrelationID:(nullable SDLRPCRequest *)request highestCorrelationIDReceived:(NSInteger)highestCorrelationIDReceived {
```

Comment:
1. I think this can be a class method
2. `BOOL` is the correct Boolean type for Obj-C
3. This shouldn't have a nullable `request`, make sure it exists before calling this method instead?